File: rd-net/RdFramework/Base/ISingleKeyProtocolContextHandler.cs
using JetBrains.Rd.Impl;
using JetBrains.Serialization;

namespace JetBrains.Rd.Base
{
  internal interface ISingleKeyProtocolContextHandler
  {
    void ReadValueAndPush(SerializationCtx context, UnsafeReader reader);
    void PopValue();
    
    void WriteValue(SerializationCtx context, UnsafeWriter writer);
  }
  
  internal interface ISingleKeyProtocolContextHandler<T> : ISingleKeyProtocolContextHandler
  {
    RdContextKey<T> Key { get; }
    ContextValueTransformer<T> ValueTransformer { get; set; }

    T ReadValue(SerializationCtx context, UnsafeReader reader);
  }
}

File: rd-kt/rd-framework/src/commonMain/kotlin/com/jetbrains/rd/framework/base/ISingleKeyProtocolContextHandler.kt
package com.jetbrains.rd.framework.base

import com.jetbrains.rd.framework.AbstractBuffer
import com.jetbrains.rd.framework.RdContextKey
import com.jetbrains.rd.framework.SerializationCtx
import com.jetbrains.rd.framework.impl.ContextValueTransformer
import com.jetbrains.rd.framework.impl.ContextValueTransformerDirection

internal interface ISingleKeyProtocolContextHandler<T: Any> {
    val key: RdContextKey<T>

    var myValueTransformer : ContextValueTransformer<T>?

    @Suppress("UNCHECKED_CAST")
    fun writeValue(ctx: SerializationCtx, writer: AbstractBuffer)

    fun readValue(ctx: SerializationCtx, reader: AbstractBuffer): T?

    @Suppress("UNCHECKED_CAST")
    fun getValueTransformed(): T? {
        return transformValueToProtocol(key.value)
    }

    fun transformValueToProtocol(valueRaw: T?): T? {
        val transformer = myValueTransformer
        return if(transformer != null) transformer(valueRaw, ContextValueTransformerDirection.WriteToProtocol) else valueRaw
    }

    fun transformValueFromProtocol(value: T?): T? {
        val transformer = myValueTransformer
        return if(transformer != null)
            transformer(value, ContextValueTransformerDirection.ReadFromProtocol)
        else
            value
    }
}
